### PR TITLE
add diagnostic page with lambda expressions warnings

### LINKED ISSUE #34655: [C# 12]: New Feature - Lambda default parameters
### Visual Studio release

Other (please put exact version in description textbox)

### Link to speclet

- [Speclet](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-method-group-defaults.md)

### Link to Pull request

- [PR](https://github.com/dotnet/roslyn/pull/65438)

### New behavior

Introduced in Visual Studio 17.5p2.

Lambda expressions can now include both default parameters and `params` arrays as parameters.

Method group conversions can also include default parameters or params arrays.

### Areas affected

- [X] **Add to What's new in C# 12**: Add links to updated docs in the What's new in C# 12 article.
- [X] **Update language reference**: Detail the areas in the language reference that will be updated for this feature.
- [x] **Update compiler messages**: Add new (or updated) pages for new compiler messages. Include related message.
- [x] ~~*Include in *Everyday C#* section**: Is this new feature included in Everyday C#?~~
- [x] ~~**Tutorials**: Are new tutorials required for this feature? If so, detail below~~

### Language reference tasks

- Lambda expression declarations are affected.
- Invoking lambda expressions are affected.

### Tutorial tasks

- N/A

### Everyday C# tasks

- N/A

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #34655

First commit: The feature Default parameters for lambda expressions adds three new errors/warnings. Add a page for those three: *CS9098*, *CS9099*, *CS9100*

Second commit: Move content from several existing warning and error pages to this page: *CS0748*, *CS0834*, *CS0835*, *CS1621*, *CS1628*, *CS1632*, *CS1673*, *CS1686*, *CS1706*, *CS1951*, *CS1952*, *CS8153*, *CS8155*, *CS8175*.  Check for all open issues on any of these pages. There were none.

In addition, look for existing warnings related to lambda expression declarations that aren't covered. Add notes for these errors and warnings: *CS1989*, *CS2037*, *CS8030*, *CS8072*, *CS8074*, *CS8075*, *CS8916*, *CS8971*, *CS8972*, *CS8975*

Finally, grammar check and fix up links.

You can review the entire PR as one change, or review commit-by-commit to see the changes incrementally.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md](https://github.com/dotnet/docs/blob/4300fc5de37b8c2fc9815716db261a1dfe384b93/docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md) | [docs/csharp/language-reference/compiler-messages/lambda-expression-errors](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/lambda-expression-errors?branch=pr-en-us-35260) |
| [docs/csharp/language-reference/operators/lambda-expressions.md](https://github.com/dotnet/docs/blob/4300fc5de37b8c2fc9815716db261a1dfe384b93/docs/csharp/language-reference/operators/lambda-expressions.md) | [Lambda expressions and anonymous functions](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions?branch=pr-en-us-35260) |
| [docs/csharp/misc/cs0837.md](https://github.com/dotnet/docs/blob/4300fc5de37b8c2fc9815716db261a1dfe384b93/docs/csharp/misc/cs0837.md) | [docs/csharp/misc/cs0837](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0837?branch=pr-en-us-35260) |
| [docs/csharp/programming-guide/classes-and-structs/local-functions.md](https://github.com/dotnet/docs/blob/4300fc5de37b8c2fc9815716db261a1dfe384b93/docs/csharp/programming-guide/classes-and-structs/local-functions.md) | [Local functions (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions?branch=pr-en-us-35260) |

### CODE REVIEW

**Comment by BillWagner** [comment]
Preview links for articles modified before fixing links:

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md](https://github.com/dotnet/docs/blob/0abd4925bcde29411a5d3f2a787c8046478bfe67/docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md) | [docs/csharp/language-reference/compiler-messages/lambda-expression-errors](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/lambda-expression-errors?branch=pr-en-us-35260) |
| [docs/csharp/misc/cs0837.md](https://github.com/dotnet/docs/blob/0abd4925bcde29411a5d3f2a787c8046478bfe67/docs/csharp/misc/cs0837.md) | [Compiler Error CS0837](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0837?branch=pr-en-us-35260) |

**Comment by BillWagner** [comment]
I *so* wanted to keep this text, but it didn't flow well in the combined article:

image

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+```
+
+## Conversion to expression trees
+
+Many legal lambda expressions can't be converted into the data structures supported by an expression tree. Attempting to use any of the constructs not supported by expression trees causes one of the following errors:
+
+- **CS0834** - *A lambda expression must have an expression body to be converted to an expression tree.*
+- **CS0835** - *Cannot convert lambda to an expression tree whose type argument 'type' is not a delegate type.*
+- **CS1951** - *An expression tree lambda may not contain an `in`, `out`, or `ref` parameter.*
+- **CS1952** - **An expression tree lambda may not contain a method with variable arguments.**
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:175]: Any reason this one is bolded?

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+- **CS1989** - *Async lambda expressions cannot be converted to expression trees.*
+- **CS2037** - *An expression tree lambda may not contain a COM call with ref omitted on arguments.*
+- **CS8072** - *An expression tree lambda may not contain a null propagating operator.*
+- **CS8074** - *An expression tree lambda may not contain a dictionary initializer.*
+- **CS8075** - *An extension Add method is not supported for a collection initializer in an expression lambda.*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:180]:
```suggestion
- **CS8075** - *An extension `Add` method is not supported for a collection initializer in an expression lambda.*
```

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+- **CS8153** - *An expression tree lambda may not contain a call to a method, property, or indexer that returns by reference.*
+- **CS8155** - *Lambda expressions that return by reference cannot be converted to expression trees.*
+- **CS8972** - *A lambda expression with attributes cannot be converted to an expression tree.*
+
+The general limitations on expressions tree are:
+
+- Attributes can
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:187]: Looks like there's an unintentional content loss and line break here.

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+- t applied to the lambda expression, its parameters or return.
+- Statement lambdas aren't allowed. It must be an expression lambda.
+- The [null propagating operator](../operators/member-access-operators.md#null-conditional-operators--and-) isn't allowed.
+- [Dictionary initializers](../../programming-guide/classes-and-structs/object-and-collection-initializers.md#collection-initializers) aren't allowed. Neither can extension add methods.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:191]:
```suggestion
- [Dictionary initializers](../../programming-guide/classes-and-structs/object-and-collection-initializers.md#collection-initializers) aren't allowed. Neither are extension `Add` methods.
```

File: docs/csharp/programming-guide/classes-and-structs/local-functions.md [hunk 5]
```diff
@@ -149,7 +149,7 @@ One final advantage not demonstrated in this sample is that local functions can
 
 :::code language="csharp" source="snippets/local-functions/Program.cs" id="YieldReturn" :::
 
-The `yield return` statement is not allowed in lambda expressions, see [compiler error CS1621](../../misc/cs1621.md).
+The `yield return` statement is not allowed in lambda expressions, see [compiler error CS1621](../../language-reference/compiler-messages/lambda-expression-errors.md#syntax-limitations-in-lambda-expressions).
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/local-functions.md:152]:
```suggestion
The `yield return` statement is not allowed in lambda expressions. For more information, see [compiler error CS1621](../../language-reference/compiler-messages/lambda-expression-errors.md#syntax-limitations-in-lambda-expressions).
```

**Review by gewarren — APPROVED** [review]
Nice work!

**Review by gewarren — APPROVED** [review]
Nice work!

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md (hunk 1 above)
Comment (BillWagner) [inline docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:175]:
```suggestion
- **CS1952** - *An expression tree lambda may not contain a method with variable arguments.*
```

File: docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+- **CS8175**: *Cannot use ref local inside an anonymous method, lambda expression, or query expression.*
+
+All the following constructs are disallowed in lambda expressions:
+
+- `yield` statements (`yield return` or `yield break`)
+- call a method using `in`, `ref` or `out` parameters
+- `ref` local variables
+- `break`, `goto` or `continue` statements
+- `this` access when `this` is a `struct` type
+- Anonymous methods or lambda expressions inside another expression, such as an Attribute constructor.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-messages/lambda-expression-errors.md:119]: Nope. None are complete sentences.

**Pull request merged**